### PR TITLE
Add documentation for getData(), spawnObjectData(), and using these to emulate "Create States"

### PR DESCRIPTION
**Pull request by RedKrieg**
These functions are mentioned briefly in the 12.4.2 patch notes: https://www.tabletopsimulator.com/news/patch-notes/v12-4-0-line-tool-revamp-ghost-preview-tags-physics-improvements-and-more

They work similarly to getJSON and spawnObjectJSON().

I also corrected an issue with the spawnObjectJSON() example, since it previously would cause an infinite spawn loop due to using onLoad() to clone self.  My solution was to use onObjectDrop() instead.

### CODE REVIEW

File: docs/base.md [hunk 1]
```diff
@@ -212,8 +213,50 @@ end
...
+####spawnObjectData(...)
 
+Spawns an Object using a data table. Works with [getData()](object.md#getdata). It works just like spawnObject, but instead of a `type`, you supply a `data` table. The other parameters will overwrite those in the Data.
 
+!!!tip
+	Spawned Objects take a moment to be physically spawned into the game. The purpose of the callback functionality is to allow you to run additional actions after the Object has been initiated fully into the instance. You can also add a delay after spawning using a [Wait](wait.md) function.
+
+!!!info "spawnObjectData(parameters)"
+	* [<span class="tag tab"></span>](types.md) **parameters**: A Table of parameters used to determine how spawnObjectData will act.
+		* [<span class="tag tab"></span>](types.md) **parameters.data**: [getData()](object.md#getdata) string.
```
Comment (Bone-White) [inline docs/base.md:225]: Probably need to remove string from the end of this line; possibly replace with table?

Comment (RedKrieg) [reply docs/base.md]: You're right, good catch.  I've added a new commit to address that, thanks.

**Review by JoshDM — COMMENTED** [review]
These look good from a presentation standpoint, but I don't know enough of the truthiness of the actual api code to truly approve.

File: docs/base.md [hunk 2]
```diff
@@ -212,8 +213,79 @@ end
...
+
+``` Lua
+function onObjectDrop(player_color, dropped_object)
+	futureName = "Spawned By Script!"
+	spawnParams = {
+		data              = dropped_object.getData(),
+		position          = {x=0, y=3, z=-5},
+		rotation          = {x=0, y=90, z=0},
+		scale             = {x=2, y=2, z=2},
+		sound             = false,
```
Comment (Benjamin-Dobell) [inline docs/base.md:244]: Could you please remove `sound` and `snap_to_grid`. _Technically_ TTS will accept those options, however I'm pretty sure they don't do anything.

File: docs/base.md [hunk 3]
```diff
@@ -212,8 +213,79 @@ end
...
+function spawn_callback(object_spawned, name, color)
+	object_spawned.setName(name)
+	object_spawned.setColorTint(color)
+end
+```
+
+!!!tip
+    You can modify an object's data prior to spawning, this is currently the only way to "Create States" via API.
+
+``` Lua
```
Comment (Benjamin-Dobell) [inline docs/base.md:262]: This is a nice little snippet. However, it's not really an "example" in that it's doing something very specific/opinionated that doesn't necessarily pertain to API itself i.e. it's quite a bit more than just calling the documented function.

These sorts of complicated examples can confuses users who are trying to do something much simpler or different than the example.

Could it please be removed? The tip above though is still useful though, so I think it's fine to remain.

File: docs/base.md [hunk 4]
```diff
@@ -236,10 +308,10 @@ Spawns an Object using a JSON string. Works with [getJSON()](object.md#getjson).
 			* {>>A reference to the object spawned is always passed to callback_function. See the example for how to access it.<<}
 
 ``` Lua
-function onLoad()
+function onObjectDrop(player_color, dropped_object)
```
Comment (Benjamin-Dobell) [inline docs/base.md:311]: This should be a separate PR.

Also, whilst I understand the motivation for the change, I'd actually rather the example just call the function and not confuse users with potentially unrelated. Newbies can get confused and think the spawning _must_ happen in these callbacks. Best to keep things short and sweet. So if you do submit a PR for this, I reckon just remove the calling context altogether.

**Review by Benjamin-Dobell — CHANGES_REQUESTED** [review]

File: docs/base.md [hunk 5]
```diff
@@ -212,8 +213,79 @@ end
 
 ---
 
+####spawnObjectData(...)
 
+Spawns an Object using a data table. Works with [getData()](object.md#getdata). It works just like spawnObject, but instead of a `type`, you supply a `data` table. The other parameters will overwrite those in the Data.
 
+!!!tip
```
[the remark this answers is not in the document]
Comment (RedKrieg) [reply docs/base.md:220]: Should I remove the tip for spawnObjectJSON too?  This is a copy-paste of what was already there.

File: docs/base.md (hunk 3 above)
Comment (RedKrieg) [reply docs/base.md:262]: I'll remove the snippet and leave the tip with a mention of the 'States' key added.

File: docs/base.md (hunk 4 above)
Comment (RedKrieg) [reply docs/base.md:311]: I'll change this back to onLoad for the purposes of separating the "onLoad is a dumb place to spawn a copy of yourself" issue in to a new PR.

File: docs/base.md (hunk 2 above)
Comment (RedKrieg) [reply docs/base.md]: I removed these for spawnObjectData but left them for spawnObjectJSON as you've requested changes to that be in a different pull request.

**Comment by Benjamin-Dobell** [comment]
_Ugh!_ So ah, my bad.

I did _not_ mean to close this. I was just cleaning it up locally. I'll push this manually.

**Comment by Benjamin-Dobell** [comment]
Merged with https://github.com/Berserk-Games/Tabletop-Simulator-API/commit/25fba90acd912306e7d99d2211824eb2233fe391
